Collect logs in problem report

jvff opened the pull request:
This PR changes the behavior of collecting logs, so that it is moved from the GUI to the `problem-report` binary. The binary now uses a fixed path to where the logs are stored.

Checklist for a PR:

* [x] Describe the change in **`CHANGELOG.md`**. Only applicable if the change has any impact for a user.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)
<!-- Reviewable:end -->

pinkisemils commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

faern commented:
We still need to supply logs to the binary as well, in combination with the ones it knows about. The GUI logs will be in different directories from the daemon logs on the *nix platforms (the GUI process running as a normal user can't write to `/var/log`). So the GUI will have to supply "extra logs" as arguments to the `problem-report` tool and the tool will combine those together with the logs in the daemon directory that it already knows about.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-:-LCYwZp9A6W2_REtJGnO:b-mrivgg)*
<!-- Sent from Reviewable.io -->

faern commented:
For the platforms where the log dirs just happens to be the same for the GUI and the daemon the `problem-report` tool will find the GUI logs by itself. We can avoid duplicate logs by making `problem-report` check for every file it iterates through in it's internal known directory and removes that path from the "extra logs" supplied. After including all logs in its known dir it can loop the remaining extra logs and include those.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-:-LCZ0E7y1KYhyIRT6Y7G:b-chsdqp)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 51 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbpIaj1m9cmHrrioIu:-LCbpIaj1m9cmHrrioIv:b99lr6o) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L51)):*
> ```Rust
> /// Location of log files to be collected
> #[cfg(windows)]
> fn log_directory() -> PathBuf {
> ```

Our files usually start with imports -> consts -> `error_chain!` -> code. Could we put these new functions further down to keep that format? But I see, these are basically constants, is that why you wanted them at the top? To take that further they could be defined as `lazy_static`s and be "real" constants.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 78 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbqF8n59stQYDFxPWO:-LCbqF8n59stQYDFxPWP:b-862039) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L78)):*
> ```Rust
>     errors {
>         LogDirError(path: PathBuf) {
>             description("Error listing the files in the specified log directory")
> ```

The log directory is never "specified", it's built in and automatic. Maybe ".. in the mullvad-daemon log directory", since that is basically what it is?

---

*[mullvad-daemon/src/bin/problem-report.rs, line 217 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbrHx88UV94wDkRF1z:-LCbrHx88UV94wDkRF2-:br8h1vm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L217)):*
> ```Rust
>         if file_path.extension() == log_extension && !logs_to_ignore.contains(&file_path) {
>             problem_report.add_log(&file_path);
>             logs_to_ignore.insert(file_path);
> ```

Do you really need to add the iterated files to the `logs_to_ignore`? With this type of design I don't see why you need to have a `HashSet` really. When you iterate a single folder you will never end up with duplicate entries, since a directory will not list the same file twice. It would probably be enough to check in each iteration if the file exists in `logs_to_ignore`?

When I suggested `HashSet` I thought you would first add *all* logs from the known directory and then start adding the extra logs and during that operation ignore logs that were already added in the automatic step. Then you would need the `HashSet` to know which logs the directory iteration had added :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 218 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCcTVFw5f1QBSR1eDif:-LCcZiFV3n0Qz8nfFSYe:bashhi3) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/33537f96f45ab70ae5f3cd97c3ada3414b05727e/mullvad-daemon/src/bin/problem-report.rs#L218)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

One could also just store the collected logs in a hash set or a map to ensure that only unique logs are constructed.
</blockquote></details>

We want the order to be the same as the order in which you add them to the report. Which is why a hash based collection is not suitable and we instead use a vector.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 218 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCcTVFw5f1QBSR1eDif:-LCd6K8_ANJImPbRaOY1:beofbos) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/33537f96f45ab70ae5f3cd97c3ada3414b05727e/mullvad-daemon/src/bin/problem-report.rs#L218)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

We want the order to be the same as the order in which you add them to the report. Which is why a hash based collection is not suitable and we instead use a vector.
</blockquote></details>

Perhaps using an `IndexMap` from https://crates.io/crates/indexmap would be a nice alternative?

---

*[mullvad-daemon/src/bin/problem-report.rs, line 51 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbpIaj1m9cmHrrioIu:-LCd37lI92sTrkN6Ea-t:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L51)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Our files usually start with imports -> consts -> `error_chain!` -> code. Could we put these new functions further down to keep that format? But I see, these are basically constants, is that why you wanted them at the top? To take that further they could be defined as `lazy_static`s and be "real" constants.
</blockquote></details>

Done.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 78 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbqF8n59stQYDFxPWO:-LCd3J0S2SPveyU3zJtC:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L78)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

The log directory is never "specified", it's built in and automatic. Maybe ".. in the mullvad-daemon log directory", since that is basically what it is?
</blockquote></details>

Done.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 173 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCcSRudCKoEjqQpMTQ7:-LCdB6xG7xxGPK6ZRLuY:b-ubb87t) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L173)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

These path's could be collected into a hash set from the get go.
</blockquote></details>

Good catch, thanks!

---

*[mullvad-daemon/src/bin/problem-report.rs, line 217 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbrHx88UV94wDkRF1z:-LCd4_MO5pHr_xUDjhEX:b8b6j71) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L217)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Do you really need to add the iterated files to the `logs_to_ignore`? With this type of design I don't see why you need to have a `HashSet` really. When you iterate a single folder you will never end up with duplicate entries, since a directory will not list the same file twice. It would probably be enough to check in each iteration if the file exists in `logs_to_ignore`?

When I suggested `HashSet` I thought you would first add *all* logs from the known directory and then start adding the extra logs and during that operation ignore logs that were already added in the automatic step. Then you would need the `HashSet` to know which logs the directory iteration had added :)
</blockquote></details>

You're right about not needing to add the iterated files to `logs_to_ignore`. I've removed it.

Regarding using the `HashSet` here with only the manually specified files, it was just a matter of the order of adding things. I had decided to keep all manually specified logs in order and at the end of the collected log file. Should I change this?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 2 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 218 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCcTVFw5f1QBSR1eDif:-LCdXeVR1Dhrnvj2fe5m:b73gznx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/33537f96f45ab70ae5f3cd97c3ada3414b05727e/mullvad-daemon/src/bin/problem-report.rs#L218)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Perhaps using an `IndexMap` from https://crates.io/crates/indexmap would be a nice alternative?
</blockquote></details>

Since the first `String` in the tuple is the redacted display version of the path it's not really a unique identifier for the file path. Therefore I would probably not use it as such in a map.
Which is why a `Vec<(String, String)>` paired with a `HashSet<PathBuf>` makes a good combination for holding the data and for detecting collisions.

You could avoid having to pass around the `logs_to_ignore` if you move the `HashSet<PathBuf>` to be a member of the `ProblemReport` and make it ignore duplicates internally in its `add_log` method. I suspect that might make things simpler?

---

*[mullvad-daemon/src/bin/problem-report.rs, line 217 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbrHx88UV94wDkRF1z:-LCdX-nw5j8b7QeazWcl:b-5za9ul) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L217)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

You're right about not needing to add the iterated files to `logs_to_ignore`. I've removed it.

Regarding using the `HashSet` here with only the manually specified files, it was just a matter of the order of adding things. I had decided to keep all manually specified logs in order and at the end of the collected log file. Should I change this?
</blockquote></details>

I agree the extra logs specified on the command line should end up in the report in the same order as specified and after the automatically collected ones.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 218 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCcTVFw5f1QBSR1eDif:-LCdeNrmEblDYm43NweA:bp1jwis) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/33537f96f45ab70ae5f3cd97c3ada3414b05727e/mullvad-daemon/src/bin/problem-report.rs#L218)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

It could simplify some things, but then I'd have to change the `collect_logs` function so that the manually added logs are placed before the automatically added logs. Otherwise, in case the user specifies a log that will be automatically added as well, it won't be in the correct order. Would this be acceptable?
</blockquote></details>

It's all up to what the correct order is. I like the other way around. That all automatic logs are always included in the order they are listed by the OS. Then after that comes all extra logs given that were not already included. That's how I initially imagined it. Because otherwise the `EXTRA LOGS` arguments can mess around with the automatic logs.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 218 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCcTVFw5f1QBSR1eDif:-LCdtK1E6GCxTyHNYXlW:b-nhbgrw) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/33537f96f45ab70ae5f3cd97c3ada3414b05727e/mullvad-daemon/src/bin/problem-report.rs#L218)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

It's all up to what the correct order is. I like the other way around. That all automatic logs are always included in the order they are listed by the OS. Then after that comes all extra logs given that were not already included. That's how I initially imagined it. Because otherwise the `EXTRA LOGS` arguments can mess around with the automatic logs.
</blockquote></details>

Done, either was is fine by me.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 217 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCbrHx88UV94wDkRF1z:-LCdssEp1o3WdkRUbz0O:b-tj6frx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e5a0c296260c067199f0c7824c7304b2c5e02308/mullvad-daemon/src/bin/problem-report.rs#L217)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I agree the extra logs specified on the command line should end up in the report in the same order as specified and after the automatically collected ones.
</blockquote></details>

Done, simplified the code a little. I also simplified it to use an `add_logs` method. Not sure if the `logs_from_log_directory` function became a little too cryptic.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

jvff commented:
*[mullvad-daemon/src/bin/problem-report.rs, line 218 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCcTVFw5f1QBSR1eDif:-LCdz4Sb3Q1TCSHihakW:b6lg6w9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/33537f96f45ab70ae5f3cd97c3ada3414b05727e/mullvad-daemon/src/bin/problem-report.rs#L218)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Done, either was is fine by me.
</blockquote></details>

Comment correction: `s/was//`

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 202 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCeEjfDAgj7fQstVszM:-LCeEjfDAgj7fQstVszN:b8auj83) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/b169b05175d23d23ac6a1f69887e108e25e3512b/mullvad-daemon/src/bin/problem-report.rs#L202)):*
> ```Rust
> }
> 
> fn logs_from_log_directory() -> Result<impl Iterator<Item = PathBuf>> {
> ```

Congrats on likely being the first one to commit an `impl Trait` to our app :wink: 

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

pinkisemils commented:
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 205 at r5](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCefkcl462Zx0GrQt0g:-LChpQ8P5knPzROVDRuQ:birktnh) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/23739829965122302b99cd86647d65dde553cc78/mullvad-daemon/src/bin/problem-report.rs#L205)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This return type is such a complex type for something that feels very simple. It should be a list of files, or an error. But I guess that is what one get's since `fs::read_dir` is crazy and both return `Result`s on initial call and on each iteration. I have no idea what type of errors the inner `Result` could actually be? Which OS lets you start listing a directory but then gives an error later?
</blockquote></details>

I think some systems, have "iterator-like" syscalls. An initial call is made to open the directory, and successive calls to a different syscall return the entries, one-by-one. Each syscall can fail, so I guess `read_dir` tries to be as generic as possible.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 290 at r5](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCef9Dd29oPIzlyZdjj:-LChyoTm70flsJ6bkFKN:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/23739829965122302b99cd86647d65dde553cc78/mullvad-daemon/src/bin/problem-report.rs#L290)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I feel like the `Result` should be handled one level up. Sending a result as an argument to a method feels a bit off to me. That would require `add_error` to be public, but I don't see how that would hurt. Might be useful for a consumer of `ProblemReport` to attach arbitrary error messages to it? If we go this direction one could also add a `message` argument to `add_error` that can end up in the first `String` in the `self.logs` tuple. Thus one can attach any error with any header. In that way one can describe if the error was because we failed to list a directory or if a certain element in the listing was an error etc.

IMO words like "to" can be skipped in method names in this type of context. Just `try_add_logs` would be shorter and just as understandable. Compare with `try_from`, `try_fold` etc from the standard library.
</blockquote></details>

Done.

---

*[mullvad-daemon/src/bin/problem-report.rs, line 312 at r5](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170#-LCegpzp64q_RcF9lW4q:-LChyiYe9n0Nql_f4Br-:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/23739829965122302b99cd86647d65dde553cc78/mullvad-daemon/src/bin/problem-report.rs#L312)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Chaining with `TryToAddLogError` here makes `add_error` very specific as a helper to `try_to_add_logs`. It feels like that method if any should chain the error and this one should just be responsible for adding any error to the report?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->

pinkisemils commented:
Reviewed 2 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/170)*
<!-- Sent from Reviewable.io -->
